### ISSUE TITLE
logstore GC shall be tired to historical store progress

### ISSUE THREAD

**Issue by chenziliang** [timeplus-io/proton#580]
**Describe what's wrong**

If we set a short period of retention or size for logstore of a stream, the live data can get pruned quickly even when they are not indexed to historical data store which results in data loss.

When logstore does the retention check, it need check the historical progress as well. If the historical data store progresses to sequence number N, logstore can only prune the data before N.

**How to reproduce**

**Error message and/or stacktrace**

**Additional context**

**Comment by rmg55** [comment]
This would be really great - any update on this effort?

**Comment by yokofly** [comment]
Yes, this issue has been fixed in proton:3.0.1https://github.com/timeplus-io/proton/releases/tag/v3.0.1


  The fix:
  - When a stream starts up, it registers a callback so the logstore can query the historical store's committed sequence number ([StreamShardStore.cpp](https://github.com/timeplus-io/proton/blob/b506213967df087a913f6c2e1a17badff4dfc6f3/src/Storages/Stream/StreamShardStore.cpp#L132-L141) )
  - During retention checks, the logstore calls this callback to get the applied sequence from historical storage ([Log.h](https://github.com/timeplus-io/proton/blob/b506213967df087a913f6c2e1a17badff4dfc6f3/src/Cluster/LocalLog/Log/Log.h#L258-L264))
  - Segments are only deleted if their base sequence is less than the historical store's applied sequence:
https://github.com/timeplus-io/proton/blob/b506213967df087a913f6c2e1a17badff4dfc6f3/src/Cluster/LocalLog/Log/Loglet.cpp#L920-L948


  Result: Logstore data is only pruned after it's been successfully indexed to the historical store, preventing data loss even with short retention periods.
